r_gl.cpp

```diff
@@ -614,13 +963,67 @@ void GLRender::RenderColliders(Camera* camera, HKD_Model** models, uint32_t numM
 
 void GLRender::RenderEnd(void)
 {
+
+    // At this point the GL default FBO must be active!
+    glBindFramebuffer(GL_FRAMEBUFFER, 0);
+
+    glPolygonMode(GL_FRONT_AND_BACK, GL_FILL);
+    glEnable(GL_DEPTH_TEST);
+    glEnable(GL_CULL_FACE);
+    glFrontFace(GL_CCW);
+    glCullFace(GL_BACK);
+    glDepthFunc(GL_LESS);
+    glEnable(GL_BLEND);
+    glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
+    
+    // FIX: Some buffer *HAS* to be bound apparently before a draw call 
+    // so OpenGL doesn't freak out... So this buffer is not even used 
+    // duing drawing because the compositing step is completely GPU driven.
+    m_FontBatch->Bind();
+
+    // Composite all the FBOs together
+    m_CompositeShader->Activate();
+
+    GLuint texLoc3d = glGetUniformLocation(m_CompositeShader->Program(),
+                                           "main3dSceneTexture");
+
+    GLuint texLoc2d = glGetUniformLocation(m_CompositeShader->Program(),
+                                           "screenspace2dTexture");
+    
+    glUniform1i(texLoc3d, 0);
+    // Bind the 3d scene FBO and draw it.
+    CglRenderTexture main3dSceneTexture = m_3dFBO->m_ColorTexture;
+    glActiveTexture( GL_TEXTURE0 );
+    glBindTexture( GL_TEXTURE_2D, main3dSceneTexture.m_gl_Handle );
+     
+    glUniform1i(texLoc2d, 1);
+    // Bind the 2d screenspace FBO texture and draw on top.
+    CglRenderTexture screenSpace2dTexture = m_2dFBO->m_ColorTexture;
+    glActiveTexture( GL_TEXTURE1 );
+    glBindTexture( GL_TEXTURE_2D, screenSpace2dTexture.m_gl_Handle ); 
+
+    glDrawArrays( GL_TRIANGLES, 0, 6 );
+
+    // FIX: This is the reason why textures in OpenGL suck!!! If we are
+    // not careful, and *don't* set the state back, then the following
+    // rendercommands for the next frame are all messed up due to the wrong
+    // texture unit being set from the previous call 
+    // ( glActiveTexture( GL_TEXTURE1 ); It is very easy to forget
+    // (I just did!). Luckily, we can use bindless
+    // textures in OpenGL which we *definitely* want to use in a
+    // facelift of the renderer.
+    glBindTexture( GL_TEXTURE_2D, 0 );
+    glActiveTexture( GL_TEXTURE0 );
+
+    // Render ImGui Elements ontop of everything.
     ImGui::Render();
     ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
 
     SDL_GL_SwapWindow(m_Window);
 
     m_ImPrimitiveBatch->Reset();
     m_ImPrimitiveBatchIndexed->Reset();
+    m_FontBatch->Reset();
```

Comment:
Good catch. Actually, in *theory* even the `m_FontBatch` should not have to be reset here either because it is expected to `Flush*` both shapes and fonts earlier to `RenderEnd`. However, not resetting them here could potentially crash the program if the user *never* calls either `Flush*` function (it would crash with a good errormessage however that the batch has run full and no more memory is available). 
Long story short, the `Reset` function are basically free, so no harm in calling them here as well, just to be safe!
Thanks.